helxplatform/appstore
# Artillery testing

## Adds smoke and load testing for Appstore using the npm Artillery package.
### Usage
- Install using `make install.artillery`. Automatically installs as a dependency of `make install`.
- Smoke/load tests can be run using `make test.artillery`. These tests will also be run using the appstore-chart with `helm test` and configured via Helm values.
#### Environment
- `ARTILLERY_ENVIRONMENT=smoke|load` and `ARTILLERY_TARGET=<appstore_url>`should be set when using `make test.artillery`
- When running load tests, `ARTILLERY_DURATION=<seconds_to_run>` and `ARTILLERY_ARRIVAL_RATE=<arrivals_per_second>` should also be set to specify load duration and intensity.
- If any required env configuration is unset when running the command, it will convey an appropriate error message.
### Design
#### Smoke vs load tests
Smoke tests are somewhat analogous to regular unit tests. It will simulate a single user and ensure that important functionalities of the site are working properly. Essentially, the smoke testing aspect ensures that Appstore is still deploying properly after making changes and that everything is still in order.

Load tests will simulate a large amount of users simultaneously, not only to ensure that key functionalities are working (as is done with smoke tests), but also to ensure that Appstore is able to handle a somewhat "realistic" flow of traffic without buckling (e.g. timing out, erroring, etc.).

The tests are designed such that both smoke and load testing follows the same workflow. The only difference is that when load testing is enabled, many users will be instantiated rather than just one.
#### Tests
The current flow of the tests is as follows:
- Log in using a random testing account.
  - Note that currently, if testing locally rather than through helm, then test users will need to be created by setting `CREATE_TEST_USERS=true` and `TEST_USERS_PATH=artillery-tests/payloads` before running Appstore with `make start`.
- Retrieve the currently running apps on the account.
- Retrieve apps that can be created by the user for the current deployment of Appstore.
- Choose a random one of these apps, and create it with minimum resources.
- Confirm that the new app was created for the user.
- Delete the new app, and confirm it was deleted.
- Log out of the account.

A couple of current drawbacks presented by this test design:
- Users never actually interact with the instantiated apps. They simply spawn them, confirm they are created, wait, and then delete them.
-  Similarly, since apps are deleted relatively quickly to keep testing at least somewhat concise, the total number of apps running at any given time isn't as high as it would naturally be if, for example, 100 students all began instances of Jupyter and used them throughout a lecture.

### Related changes
Some changes apart from the artillery testing itself were adde in order for the tests to work.
- Node/npm/curl added to the Docker container. Curl is used to ensure that the correct version of Node is installed (see Dockerfile).
- Numerous changes to the Makefile for running Artillery.
- The API endpoint `POST /api/v1/instances/`, which creates a new app, now returns the `sid` of the created app.
   - This would not normally be necessary, since you could infer the created app by looking at the apps before it was created and after, but due to limitations of Artillery, it is possible that numerous virtual users can end up logging in to the same account at once at create apps at the same time.
